### ISSUE TITLE
struct_pack: Evaluation depth exceeds 512

### ISSUE THREAD

**Issue by Priyadharshini-Prabhakaran** [alibaba/yalantinglibs#229]
### Search before asking

- [x] I searched the [issues](https://github.com/alibaba/yalantinglibs/issues) and found no similar issues.

### What happened + What you expected to happen
I created a user defined structure to be sent as a return to an RPC method call. While trying that the following error occurred

```
/home/sysadmin/my_project/yalantinglibs/include/struct_pack/struct_pack/struct_pack_impl.hpp:829:37: error: ‘constexpr’ evaluation depth exceeds maximum of 512 (use ‘-fconstexpr-depth=’ to increase the maximum)
  829 |   constexpr auto id = get_type_id<T>();
      |                       ~~~~~~~~~~~~~~^~
```

### Reproduction way
Steps to reproduce,
1. Create a user defined structure with a list as a member of the same
2. On an RPC method call from client serialize the user defined structure and send it as a return to the call
3. Compile the application
4. The above mentioned error will be reproduced

### Anything else
1) As mentioned in the error statement, I tried to set the ‘-fconstexpr-depth=’ flag to higher values in CMake files,
`set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fconstexpr-depth=4096")`
2) Is there a maximum limit for this parameter ‘-fconstexpr-depth=’?
3) Is it configurable via code as #defines?
4) I also observed few of these parameters hardcoded in the header file, was it done intentionally with a limit? For example,
```
In /home/sysadmin/my_project/yalantinglibs/include/struct_pack/struct_pack/reflection.hpp - line number: 338 - MaxVisitMembers is hardcoded to 64
```

### Are you willing to submit a PR?

- [x ] Yes I am willing to submit a PR!

**Comment by qicosmos** [comment]
could you give us your user defined structure? give the whole test code is better if possible.

**Comment by poor-circle** [comment]
I afraid that your struct cause a infinite recursion in compile time calculation. Could you show me your struct?

**Comment by Priyadharshini-Prabhakaran** [comment]
Yes the struct as a recursive member within it. A sample user defined structure is as follows,

```
struct Node
{
	int Type;					
	bool IsArray;
	std::list<uint32_t> Dimensions;
	bool IsStructure;
	std::string TypeName;
	uint16_t Index;
	bool BoolValue;			
	std::string StringValue;
 	float FloatValue;		
	double DoubleValue;
	int16_t Int16Value;		
	int32_t Int32Value;		
	int64_t Int64Value;				
	uint16_t UInt16Value;	
	uint32_t UInt32Value;	
	uint64_t UInt64Value;	
	std::list<Node> Values;    // Recursive member
};
```

I am in need of an implementation to have this recursive structure. Is there any solution for that?

**Comment by poor-circle** [comment]
> Yes the struct as a recursive member within it. A sample user defined structure is as follows,
> 
> ```
> struct Node
> {
> 	int Type;					
> 	bool IsArray;
> 	std::list<uint32_t> Dimensions;
> 	bool IsStructure;
> 	std::string TypeName;
> 	uint16_t Index;
> 	bool BoolValue;			
> 	std::string StringValue;
>  	float FloatValue;		
> 	double DoubleValue;
> 	int16_t Int16Value;		
> 	int32_t Int32Value;		
> 	int64_t Int64Value;				
> 	uint16_t UInt16Value;	
> 	uint32_t UInt32Value;	
> 	uint64_t UInt64Value;	
> 	std::list<Node> Values;    // Recursive member
> };
> ```
> 
> I am in need of an implementation to have this recursive structure. Is there any solution for that?

OK, I will try to fix that.

**Comment by poor-circle** [comment]
Fixed by #230

**Comment by Priyadharshini-Prabhakaran** [comment]
Thank you for the fix and the suggestion. I will try the same out and will keep you posted.
